### PR TITLE
[PyCDE] Move to `msft.module`/`msft.instance` and move generation into Python

### PR DESCRIPTION
**Pull request by teqdruid**
This is a _massive_ refactor. There's still some rough edges, but it should be far simpler, more logical, and "safer" from MLIR operation memory perspective. This refactor does break the instance/appid features, which I'll fix later on. That test is marked XFAIL for the moment. This closes out #1755.

I apologize for the size, but it really wasn't feasible to break it up since all this stuff is so intertwined. I also moved code and modified it slightly, which you can't see in the diff. This PR is a net code reduction if you consider: (1) the extra documentation and (2) that this'll allow me to rip out all the problematic C++ generation stuff since this PR replaces it.

### CODE REVIEW

File: frontends/PyCDE/src/pycde/module.py [hunk 1]
```diff
@@ -45,45 +47,217 @@ class Input(ModuleDecl):
...
+  """The "module" cache is specifically for parameterized modules. It maps the
+  module parameterization function AND parameter values to the class which was
+  generated by a previous call to said module parameterization function."""
+  if not isinstance(params, mlir.ir.DictAttr):
+    params = var_to_attribute(params)
+  return (func, mlir.ir.Attribute(params))
+
+
+# A memoization table for module parameterization function calls.
+_module_cache: typing.Dict[Tuple[builtins.function, mlir.ir.DictAttr],
```
Comment (mikeurbach) [inline frontends/PyCDE/src/pycde/module.py:242]: maybe put this global variable at the top of the file and name it like MODULE_CACHE?

File: frontends/PyCDE/src/pycde/support.py [hunk 2]
```diff
@@ -52,6 +52,14 @@ def get_user_loc() -> ir.Location:
   return ir.Location.unknown()
 
 
+def create_const_zero(type: ir.Type):
+  """Create a 'default' constant value of zero. Used for creating dummy values
+  to connect to extern modules with input ports we want to ignore."""
+  width = hw.get_bitwidth(type)
+  zero = hw.ConstantOp.create(ir.IntegerType.get_signless(width), 0)
+  return hw.BitcastOp(type, zero.result)
```
Comment (mikeurbach) [inline frontends/PyCDE/src/pycde/support.py:60]: nifty!

File: frontends/PyCDE/src/pycde/system.py [hunk 3]
```diff
@@ -15,49 +14,80 @@
...
 import sys
 import typing
 
+_current_system = ContextVar("current_pycde_system")
+
 
 class System:
+  """The 'System' contains the user's design and some private bookkeeping. On
+  construction, specify a list of 'root' modules which you wish to generate.
+  Upon generation, the design will be fleshed out and all the dependant modules
```
Comment (mikeurbach) [inline frontends/PyCDE/src/pycde/system.py:27]: spelling nit: i think it's `dependent`

File: frontends/PyCDE/src/pycde/system.py [hunk 4]
```diff
@@ -72,44 +102,35 @@ def graph(self, short_names=True):
...
+    """Fully generate the system unless iters is specified. Iters specifies the
+    number of generators to run. Useful for debugging. Maybe."""
+    i = 0
+    with self:
+      while len(self._generate_queue) > 0 and (iters is None or i < iters):
+        m = self._generate_queue.pop()
+        m.generate()
+        i += 1
+
+  # Broken ATM
```
Comment (mikeurbach) [inline frontends/PyCDE/src/pycde/system.py:115]: Worth raising some sort of Exception in case someone tries to use it?

**Review by mikeurbach — APPROVED** [review]
The description and high-level goals make sense. The comments and general re-structure make sense. I didn't read too far into how all the pieces fit together now, but the fact that this is a PyCDE only change, and the test updates are reasonable is great. I need to review a little more deeply, but feel free to merge and I can raise an follow up concerns I think of.

**Comment by teqdruid** [comment]
Thanks for the review @mikeurbach ! It did actually involve some additions to the python bindings, but I committed those separately as they were separable.

**Pull request merged**